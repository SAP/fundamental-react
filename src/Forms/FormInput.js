import classnamesBind from 'classnames/bind';
import { FORM_MESSAGE_TYPES } from '../utils/constants';
import FormValidationOverlay from './_FormValidationOverlay';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/input.css';

const classnames = classnamesBind.bind(styles);
/** A **FormInput** is used to collect data from the user. When a field is required,
the \`required\` property will include an asterisk (*). */
const FormInput = React.forwardRef(({
    className,
    cssNamespace,
    compact,
    disabled,
    name,
    placeholder,
    readOnly,
    type,
    value,
    validationOverlayProps,
    validationState,
    ...props
}, ref) => {

    const formInputClasses = classnames(
        `${cssNamespace}-input`,
        {
            [`${cssNamespace}-input--compact`]: !!compact,
            [`is-${validationState?.state}`]: validationState?.state
        },
        className
    );

    const formInput = (
        <input
            {...props}
            className={formInputClasses}
            disabled={disabled}
            name={name}
            placeholder={placeholder}
            readOnly={readOnly}
            ref={ref}
            type={type}
            value={value} />
    );

    return validationState ? (
        <FormValidationOverlay
            {...validationOverlayProps}
            control={formInput}
            validationState={validationState} />
    ) : formInput;
});

FormInput.displayName = 'FormInput';

FormInput.propTypes = {
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to enable compact mode */
    compact: PropTypes.bool,
    /** Set to **true** to mark component as disabled and make it non-interactive */
    disabled: PropTypes.bool,
    /** Value for the `name` attribute on the input */
    name: PropTypes.string,
    /** Localized placeholder text of the input */
    placeholder: PropTypes.string,
    /** Set to **true** to mark component as readonly */
    readOnly: PropTypes.bool,
    /** Value for the `type` attribute on the input */
    type: PropTypes.string,
    /** Additional props to be spread to the ValidationOverlay */
    validationOverlayProps: PropTypes.shape({
        /** Additional classes to apply to validation popover's outermost `<div>` element  */
        className: PropTypes.string,
        /** Additional props to be spread to the ValdiationOverlay's FormMessage component */
        formMessageProps: PropTypes.object,
        /** Additional classes to apply to validation popover's popper child `<div>` wrapping the provided children  */
        innerRefClassName: PropTypes.string,
        /** Additional classes to apply to validation popover's popper `<div>` element  */
        popperClassName: PropTypes.string,
        /** CSS class(es) to add to the ValidationOverlay's reference `<div>` element */
        referenceClassName: PropTypes.string,
        /** Additional props to be spread to the popover's outermost `<div>` element */
        wrapperProps: PropTypes.object
    }),
    /** An object identifying a validation message.  The object will include properties for `state` and `text`; _e.g._, \`{ state: \'warning\', text: \'This is your last warning\' }\` */
    validationState: PropTypes.shape({
        /** State of validation: 'error', 'warning', 'information', 'success' */
        state: PropTypes.oneOf(FORM_MESSAGE_TYPES),
        /** Text of the validation message */
        text: PropTypes.string
    }),
    /** Value for the `value` attribute on the input */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

FormInput.defaultProps = {
    type: 'text'
};

export default withStyles(FormInput);
