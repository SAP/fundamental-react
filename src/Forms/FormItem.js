import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/form-item.css';

const classnames = classnamesBind.bind(styles);

const FormItem = React.forwardRef(({ isHorizontal, isInline, children, className, cssNamespace, ...props }, ref) => {

    const formItemClasses = classnames(
        `${cssNamespace}-form-item`,
        {
            [`${cssNamespace}-form-item--inline`]: isInline,
            [`${cssNamespace}-form-item--horizontal`]: isHorizontal
        },
        className
    );

    return (
        <div
            {...props}
            className={formItemClasses}
            ref={ref}>
            {children}
        </div>
    );
});
FormItem.displayName = 'FormItem';

FormItem.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to display items in a row */
    isHorizontal: PropTypes.bool,
    /** Internal use only */
    isInline: PropTypes.bool
};

export default withStyles(FormItem);
