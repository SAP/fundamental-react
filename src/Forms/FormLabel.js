import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/form-label.css';

const classnames = classnamesBind.bind(styles);

/** A **FormLabel** is used to identify form components such as
 * **FormInput**, **Checkbox** and **TextArea**. Best practice is to use this component as a child of **FormGroup**. */
const FormLabel = React.forwardRef(({
    required,
    children,
    className,
    cssNamespace,
    disabled,
    isInlineHelp,
    includeColon,
    ...props
}, ref) => {

    const formLabelClasses = classnames(
        `${cssNamespace}-form-label`,
        {
            'is-disabled': disabled,
            [`${cssNamespace}-form-label--inline-help`]: isInlineHelp,
            [`${cssNamespace}-form-label--required`]: required,
            [`${cssNamespace}-form-label--colon`]: includeColon
        },
        className
    );

    return (
        <label
            {...props}
            aria-required={required}
            className={formLabelClasses}
            ref={ref}>
            {children}
        </label>
    );
});

FormLabel.displayName = 'FormLabel';

FormLabel.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to mark component as disabled and make it non-interactive */
    disabled: PropTypes.bool,
    /** Set to **true** to add the `:` character at the end of a label as pseudo element.*/
    includeColon: PropTypes.bool,
    /** Set to **true** if child is InlineHelp component */
    isInlineHelp: PropTypes.bool,
    /** Set to **true** for required input fields */
    required: PropTypes.bool
};

export default withStyles(FormLabel);
