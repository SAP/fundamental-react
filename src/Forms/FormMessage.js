import classnames from 'classnames';
import { FORM_MESSAGE_TYPES } from '../utils/constants';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/WithStyles/WithStyles';

const FormMessage = React.forwardRef(({ type, children, className, disableStyles, ...props }, ref) => {
    const formMessageClasses = classnames(
        'fd-form-message',
        {
            [`fd-form-message--${type}`]: !!type
        },
        className
    );
    return (
        <span
            {...props}
            className={formMessageClasses}
            ref={ref}>
            {children}
        </span>
    );
});

FormMessage.displayName = 'FormMessage';

FormMessage.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    customStyles: PropTypes.object,
    disableStyles: PropTypes.bool,
    type: PropTypes.oneOf(FORM_MESSAGE_TYPES)
};

export { FormMessage as __FormMessage };

export default withStyles(FormMessage, { cssFile: 'form-message', fonts: true, icons: true });
