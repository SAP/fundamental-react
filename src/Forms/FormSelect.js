import classnames from 'classnames';
import { FORM_STATES } from '../utils/constants';
import PropTypes from 'prop-types';
import React, { useEffect } from 'react';

const FormSelect = React.forwardRef(({ disabled, children, className, compact, disableStyles, state, ...props }, ref) => {

    useEffect(() => {
        if (!disableStyles) {
            require('fundamental-styles/dist/form-select.css');
        }
    }, []);

    const formSelectClasses = classnames(
        'fd-form-select',
        { [`is-${state}`]: state,
            'fd-form-select--compact': compact },
        className
    );

    return (
        <select
            {...props}
            className={formSelectClasses}
            disabled={disabled}
            ref={ref}>
            {children}
        </select>
    );
});

FormSelect.displayName = 'FormSelect';

FormSelect.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    compact: PropTypes.bool,
    disabled: PropTypes.bool,
    disableStyles: PropTypes.bool,
    state: PropTypes.oneOf(FORM_STATES)
};

export default FormSelect;
