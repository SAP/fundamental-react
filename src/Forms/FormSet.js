import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/WithStyles/WithStyles';

const FormSet = React.forwardRef(({ children, className, disableStyles, ...props }, ref) => {
    const formSetClasses = classnames(
        'fd-form-group',
        className
    );

    return (
        <div {...props} className={formSetClasses}
            ref={ref}>
            {children}
        </div>
    );
});

FormSet.displayName = 'FormSet';

FormSet.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    customStyles: PropTypes.object,
    disableStyles: PropTypes.bool
};

export { FormSet as __FormSet };

export default withStyles(FormSet, { cssFile: 'form-group' });
