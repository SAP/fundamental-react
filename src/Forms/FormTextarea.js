import classnamesBind from 'classnames/bind';
import CustomPropTypes from '../utils/CustomPropTypes/CustomPropTypes';
import { FORM_MESSAGE_TYPES } from '../utils/constants';
import FormValidationOverlay from './_FormValidationOverlay';
import PropTypes from 'prop-types';
import withStyles from '../utils/withStyles';
import React, { useState } from 'react';
import styles from 'fundamental-styles/dist/textarea.css';

const classnames = classnamesBind.bind(styles);

/** A **FormTextarea** is used to gather potentially lengthy input from a user.
 * Best practice is to use this component as a child of FormGroup. */
const FormTextarea = React.forwardRef(({
    className,
    compact,
    counterProps,
    cssNamespace,
    defaultValue,
    disabled,
    localizedText,
    maxLength,
    onChange,
    readOnly,
    validationOverlayProps,
    validationState,
    value,
    ...props }, ref) => {

    const getInitialCharCount = () => {
        if (typeof value === 'boolean' || value) {
            return Math.min(value.toString().length, maxLength);
        }
        if (defaultValue) {
            return Math.min(defaultValue.toString().length, maxLength);
        }
        return 0;
    };

    const hasMaxLength = maxLength >= 0;

    const [charCount, setCharCount] = useState(hasMaxLength ? getInitialCharCount() : 0);

    const getMaxLengthText = () => {
        const { charactersLeftPlural, charactersLeftSingular } = localizedText;
        let charsLeft = Math.max(maxLength - charCount, 0);
        if (charsLeft !== 1) {
            return `${charsLeft} ${charactersLeftPlural}`;
        }
        return `${charsLeft} ${charactersLeftSingular}`;
    };

    const handleOnChange = (e) => {
        if (hasMaxLength)
            setCharCount(e.target.value.length);
        if (onChange)
            onChange(e);
    };

    const formTextAreaClasses = classnames(
        `${cssNamespace}-textarea`,
        {
            [`${cssNamespace}-textarea--compact`]: compact,
            [`is-${validationState?.state}`]: validationState?.state
        },
        className
    );

    const counterClasses = classnames(
        `${cssNamespace}-textarea-counter`,
        counterProps?.className
    );

    const formTextarea = (
        <textarea
            {...props}
            className={formTextAreaClasses}
            defaultValue={defaultValue}
            disabled={disabled}
            maxLength={maxLength}
            onChange={handleOnChange}
            readOnly={readOnly}
            ref={ref}
            value={value} />
    );

    return (
        <>
            {validationState?.state ? <FormValidationOverlay
                {...validationOverlayProps}
                control={formTextarea}
                validationState={validationState} />
                : formTextarea}
            {hasMaxLength &&
                <div
                    {...counterProps}
                    className={counterClasses}>
                    {getMaxLengthText()}
                </div>
            }
        </>
    );
});

FormTextarea.displayName = 'FormTextarea';

FormTextarea.propTypes = {
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to enable compact mode */
    compact: PropTypes.bool,
    /** Props to pass to the counter div */
    counterProps: PropTypes.object,
    /** Default placeholder value for the textarea */
    defaultValue: PropTypes.string,
    /** Set to **true** to mark component as disabled and make it non-interactive */
    disabled: PropTypes.bool,
    /** Value for the `id` attribute on the element */
    id: PropTypes.string,
    /** Localized text to be updated based on location/language */
    localizedText: CustomPropTypes.i18n({
        charactersLeftPlural: PropTypes.string,
        charactersLeftSingular: PropTypes.string
    }),
    /** Set the max length of the textarea */
    maxLength: PropTypes.number,
    /** Set to **true** to mark component as readonly */
    readOnly: PropTypes.bool,
    /** Additional props to be spread to the ValidationOverlay */
    validationOverlayProps: PropTypes.shape({
        /** Additional classes to apply to validation popover's outermost `<div>` element  */
        className: PropTypes.string,
        /** Additional props to be spread to the ValdiationOverlay's FormMessage component */
        formMessageProps: PropTypes.object,
        /** Additional classes to apply to validation popover's popper child `<div>` wrapping the provided children  */
        innerRefClassName: PropTypes.string,
        /** Additional classes to apply to validation popover's popper `<div>` element  */
        popperClassName: PropTypes.string,
        /** CSS class(es) to add to the ValidationOverlay's reference `<div>` element */
        referenceClassName: PropTypes.string,
        /** Additional props to be spread to the popover's outermost `<div>` element */
        wrapperProps: PropTypes.object
    }),
    /** An object identifying a validation message.  The object will include properties for `state` and `text`; _e.g._, \`{ state: \'warning\', text: \'This is your last warning\' }\` */
    validationState: PropTypes.shape({
        /** State of validation: 'error', 'warning', 'information', 'success' */
        state: PropTypes.oneOf(FORM_MESSAGE_TYPES),
        /** Text of the validation message */
        text: PropTypes.string
    }),
    /** Value for the textarea */
    value: PropTypes.string,
    /** Callback function when the change event fires on the component */
    /**
     * Callback function; triggered when the change event fires on the HTML `<textarea>`.
     *
     * @param {SyntheticEvent} event - React's original SyntheticEvent. See https://reactjs.org/docs/events.html.
     * @returns {void}
    */
    onChange: PropTypes.func
};

FormTextarea.defaultProps = {
    localizedText: {
        charactersLeftPlural: 'characters left',
        charactersLeftSingular: 'character left'
    }
};

export default withStyles(FormTextarea);
