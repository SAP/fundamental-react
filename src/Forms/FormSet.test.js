import FormInput from './FormInput';
import FormItem from './FormItem';
import FormLabel from './FormLabel';
import FormMessage from './FormMessage';
import FormSelect from './FormSelect';
import FormSet from './FormSet';
import FormTextarea from './FormTextarea';
import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

describe('<FormSet />', () => {
    const formSet = (
        <FormSet>
            <FormItem isInline>
                <FormLabel forAttr='input-1'>Default Input</FormLabel>
                <FormInput
                    id='input-1'
                    placeholder='Field placeholder text'
                    type='text' />
            </FormItem>
            <FormItem>
                <FormLabel forAttr='input-1' required>
                    Default Input
                </FormLabel>
                <FormInput
                    id='input-1'
                    placeholder='Field placeholder text'
                    state='warning'
                    type='text' />
                <FormTextarea className='blue' id='textarea-1'>
                    Pellentesque metus lacus commodo eget justo ut rutrum varius nunc.
                </FormTextarea>
                <FormTextarea id='textarea-2'>
                    Pellentesque metus lacus commodo eget justo ut rutrum varius nunc.
                </FormTextarea>
                <FormMessage type='help'>
                    Pellentesque metus lacus commodo eget justo ut rutrum varius nunc
                </FormMessage>
                <FormMessage className='blue'>
                    Pellentesque metus lacus commodo eget justo ut rutrum varius nunc
                </FormMessage>
                <FormSelect id='select-1'>
                    <option value='1'>Duis malesuada odio volutpat elementum</option>
                    <option value='2'>Suspendisse ante ligula</option>
                    <option value='3'>Sed bibendum sapien at posuere interdum</option>
                </FormSelect>
                <FormSelect className='blue' disabled
                    id='select-1'>
                    <option value='1'>Duis malesuada odio volutpat elementum</option>
                </FormSelect>
            </FormItem>
        </FormSet>
    );

    const formSetWithClass = (
        <FormSet className='blue'>
            <FormItem isInline>
                <FormLabel forAttr='input-1'>Default Input</FormLabel>
                <FormInput
                    id='input-1'
                    placeholder='Field placeholder text'
                    type='text' />
            </FormItem>
        </FormSet>
    );

    test('create form item', () => {
        // create form set with form inputs
        let component = renderer.create(formSet);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        //create form ield set with form inputs
        component = renderer.create(formSetWithClass);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('<FormSet Prop spreading', () => {
        test('should allow props to be spread to the FormSet component', () => {
            const element = mount(<FormSet data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });

    test('forwards the ref', () => {
        let ref;
        class Test extends React.Component {
            constructor(props) {
                super(props);
                ref = React.createRef();
            }
            render = () => <FormSet ref={ref} />;
        }
        mount(<Test />);
        expect(ref.current.tagName).toEqual('DIV');
        expect(ref.current.className).toEqual('fd-form-group');
    });
});


