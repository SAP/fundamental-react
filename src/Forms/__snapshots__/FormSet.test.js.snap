// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSet /> create form item 1`] = `
<div
  className="fd-form-group"
>
  <div
    className="fd-form-item fd-form-item--horizontal"
  >
    <label
      className="fd-form-label"
      forAttr="input-1"
    >
      Default Input
    </label>
    <input
      className="fd-input"
      id="input-1"
      placeholder="Field placeholder text"
      type="text"
    />
  </div>
  <div
    className="fd-form-item"
  >
    <label
      aria-required={true}
      className="fd-form-label fd-form-label--required"
      forAttr="input-1"
    >
      Default Input
    </label>
    <div
      className="fd-popover"
      style={
        Object {
          "display": "block",
        }
      }
    >
      <div
        onFocus={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="fd-popover__control"
        >
          <input
            aria-controls="input-1"
            aria-expanded={false}
            aria-haspopup={true}
            className="fd-input is-warning"
            id="input-1"
            onClick={[Function]}
            onKeyPress={[Function]}
            placeholder="Field placeholder text"
            role="button"
            tabIndex={0}
            type="text"
          />
        </div>
      </div>
    </div>
    <textarea
      className="fd-textarea blue"
      id="textarea-1"
    >
      Pellentesque metus lacus commodo eget justo ut rutrum varius nunc.
    </textarea>
    <textarea
      className="fd-textarea"
      id="textarea-2"
    >
      Pellentesque metus lacus commodo eget justo ut rutrum varius nunc.
    </textarea>
    <select
      className="fd-form-select"
      id="select-1"
    >
      <option
        value="1"
      >
        Duis malesuada odio volutpat elementum
      </option>
      <option
        value="2"
      >
        Suspendisse ante ligula
      </option>
      <option
        value="3"
      >
        Sed bibendum sapien at posuere interdum
      </option>
    </select>
    <select
      className="fd-form-select blue"
      disabled={true}
      id="select-1"
    >
      <option
        value="1"
      >
        Duis malesuada odio volutpat elementum
      </option>
    </select>
  </div>
</div>
`;

exports[`<FormSet /> create form item 2`] = `
<div
  className="fd-form-group blue"
>
  <div
    className="fd-form-item fd-form-item--horizontal"
  >
    <label
      className="fd-form-label"
      forAttr="input-1"
    >
      Default Input
    </label>
    <input
      className="fd-input"
      id="input-1"
      placeholder="Field placeholder text"
      type="text"
    />
  </div>
</div>
`;
